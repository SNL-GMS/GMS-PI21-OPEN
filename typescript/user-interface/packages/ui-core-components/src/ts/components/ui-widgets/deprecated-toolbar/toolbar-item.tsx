/* eslint-disable react/destructuring-assignment */
import { UILogger } from '@gms/ui-util';
import React from 'react';

import type { PopoverButton } from '../popover-button';
// !FIX ESLINT CIRCULAR DEPENDENCY
// eslint-disable-next-line import/no-cycle
import * as Renderers from './toolbar-item-renderers';
import type {
  ButtonGroupItem,
  ButtonItem,
  CheckboxDropdownItem,
  CustomItem,
  DateRangePickerItem,
  DeprecatedToolbarItemBase,
  DropdownItem,
  IntervalPickerItem,
  LabelValueItem,
  LoadingSpinnerItem,
  NumericInputItem,
  PopoverItem,
  SwitchItem,
  ToolbarItem
} from './types';
import { ToolbarItemType } from './types';

const logger = UILogger.create('GMS_LOG_TOOLBAR', process.env.GMS_LOG_TOOLBAR);

export interface ToolbarItemRendererProps {
  item: ToolbarItem;
  hasIssue: boolean;
  addToPopoverMap(rank: number, ref: PopoverButton): void;
}

/**
 * Selects the correct function to render a toolbar item
 */
// eslint-disable-next-line complexity
export function ToolbarItemRenderer(props: ToolbarItemRendererProps): JSX.Element {
  const itemTypes = ToolbarItemType;
  switch ((props.item as DeprecatedToolbarItemBase).type) {
    case itemTypes.Dropdown: {
      return Renderers.renderDropdown(props.item as DropdownItem);
    }
    case itemTypes.NumericInput: {
      return Renderers.renderNumeric(props.item as NumericInputItem);
    }
    case itemTypes.IntervalPicker: {
      return Renderers.renderIntervalPicker(props.item as IntervalPickerItem);
    }
    case itemTypes.DateRangePicker: {
      return Renderers.renderDateRangePicker(props.item as DateRangePickerItem);
    }
    case itemTypes.Popover: {
      return Renderers.renderPopoverButton(props.item as PopoverItem, (key, val) =>
        props.addToPopoverMap(key, val)
      );
    }
    case itemTypes.Switch: {
      return Renderers.renderSwitch(props.item as SwitchItem);
    }
    case itemTypes.Button: {
      return Renderers.renderButton(props.item as ButtonItem);
    }
    case itemTypes.ButtonGroup: {
      return Renderers.renderButtonGroup(props.item as ButtonGroupItem);
    }
    case itemTypes.LabelValue: {
      return Renderers.renderLabelValue(props.item as LabelValueItem, props.hasIssue);
    }
    case itemTypes.CheckboxList:
      return Renderers.renderCheckboxDropdown(props.item as CheckboxDropdownItem, (key, val) =>
        props.addToPopoverMap(key, val)
      );
    case itemTypes.LoadingSpinner:
      return Renderers.renderLoadingSpinner(props.item as LoadingSpinnerItem);
    case itemTypes.CustomItem:
      return (props.item as CustomItem).element;
    default: {
      logger.warn(`default error`);
      return <div />;
    }
  }
}
