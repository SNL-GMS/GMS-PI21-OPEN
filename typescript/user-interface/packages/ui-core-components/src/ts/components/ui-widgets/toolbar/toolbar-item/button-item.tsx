import { Alignment, Button, Icon } from '@blueprintjs/core';
import React from 'react';

import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isButtonToolbarItem(object: unknown): object is ButtonToolbarItemProps {
  return (object as ButtonToolbarItemProps).onButtonClick !== undefined;
}

/**
 * Properties to pass to the {@link ButtonToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface ButtonToolbarItemProps extends ToolbarItemBase {
  onButtonClick();
  marginRight?: number;
}

/**
 * Represents a group of buttons used within a toolbar
 *
 * @param buttonItem the buttonItem to display {@link ButtonItem}
 */
// eslint-disable-next-line complexity, react/function-component-definition
export const ButtonToolbarItem: React.FC<ButtonToolbarItemProps> = ({
  marginRight = 0,
  widthPx,
  onlyShowIcon,
  style,
  disabled,
  label,
  labelRight,
  tooltip,
  icon,
  onButtonClick,
  onMouseEnter,
  onMouseOut,
  cyData
}: ButtonToolbarItemProps): ToolbarItemElement => {
  const widthAsString = widthPx ? `${widthPx}px` : undefined;
  const width = widthAsString || (onlyShowIcon ? '30px' : undefined);

  return (
    <div style={style ?? {}}>
      <Button
        className={onlyShowIcon ? 'toolbar-button--icon-only' : 'toolbar-button'}
        style={{
          width,
          marginRight: marginRight ? `${marginRight}px` : undefined
        }}
        disabled={disabled}
        alignText={onlyShowIcon && !labelRight ? Alignment.CENTER : Alignment.LEFT}
        onClick={onButtonClick}
        title={tooltip}
        onMouseEnter={onMouseEnter}
        onMouseOut={onMouseOut}
        data-cy={cyData}
      >
        {onlyShowIcon ? undefined : <span key="toolbar-button--label">{label}</span>}
        {icon ? <Icon icon={icon} title={false} /> : null}
        {!labelRight ? undefined : <span key="toolbar-button--label-right">{labelRight}</span>}
      </Button>
    </div>
  );
};
