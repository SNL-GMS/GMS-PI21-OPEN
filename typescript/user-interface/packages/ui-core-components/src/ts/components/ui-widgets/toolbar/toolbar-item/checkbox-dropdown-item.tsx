import type Immutable from 'immutable';
import React from 'react';

import { CheckboxList } from '../../checkbox-list';
import { PopoverButton } from '../../popover-button';
import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isCheckboxDropdownToolbarItem(
  object: unknown
): object is CheckboxDropdownToolbarItemProps {
  return (object as CheckboxDropdownToolbarItemProps).enumOfKeys !== undefined;
}

/**
 * props for {@link CheckboxDropdownToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface CheckboxDropdownToolbarItemProps extends ToolbarItemBase {
  /** map for checkbox values (checked/unchecked) */
  values: Immutable.Map<any, boolean>;

  /** map of background colors */
  colors?: Immutable.Map<any, string>;

  /** map of keys to display strings (labels) */
  enumKeysToDisplayStrings?: Immutable.Map<string, string>;

  enumKeysToRenderDividers?: Immutable.Map<string, boolean>;

  enumKeysToLabelStrings?: Immutable.Map<string, string>;

  /** keys */
  enumOfKeys: any;

  /** callback to change event */
  onChange(value: any): void;

  /** callback to onPopUp (list appearing) event */
  onPopUp?(ref?: HTMLDivElement): void;

  /** callback to onPopoverDismissed (list disappears) event */
  onPopoverDismissed?(): void;
}

/**
 * Represents a dropdown list of checkbox items used within a toolbar
 *
 * @param checkboxItem the checkboxItem to display {@link CheckboxDropdownItem}
 */
// eslint-disable-next-line react/function-component-definition
export const CheckboxDropdownToolbarItem: React.FC<CheckboxDropdownToolbarItemProps> = ({
  values,
  colors,
  enumKeysToDisplayStrings,
  enumKeysToLabelStrings,
  enumKeysToRenderDividers,
  enumOfKeys,
  onChange,
  onPopUp,
  onPopoverDismissed,
  style,
  label,
  tooltip,
  disabled,
  widthPx,
  popoverButtonMap,
  cyData
}: CheckboxDropdownToolbarItemProps): ToolbarItemElement => {
  const handleRef = (ref: PopoverButton, buttonMap: Map<number, PopoverButton>): void => {
    if (ref && buttonMap) {
      buttonMap.set(1, ref);
    }
  };

  return (
    <div style={style ?? {}}>
      <PopoverButton
        label={label}
        tooltip={tooltip}
        disabled={disabled}
        cyData={cyData}
        popupContent={
          <CheckboxList
            enumToCheckedMap={values}
            enumToColorMap={colors}
            checkboxEnum={enumOfKeys}
            enumKeysToDisplayStrings={enumKeysToDisplayStrings}
            enumKeysToDividerMap={enumKeysToRenderDividers}
            enumKeysToLabelMap={enumKeysToLabelStrings}
            onChange={onChange}
          />
        }
        onPopoverDismissed={onPopoverDismissed}
        widthPx={widthPx}
        onClick={onPopUp}
        ref={ref => handleRef(ref, popoverButtonMap)}
      />
    </div>
  );
};
