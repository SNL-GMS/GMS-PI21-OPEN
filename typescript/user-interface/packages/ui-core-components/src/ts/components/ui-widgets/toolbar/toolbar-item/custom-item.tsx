import type React from 'react';

import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isCustomToolbarItem(object: unknown): object is CustomItemProps {
  return (object as CustomItemProps).element !== undefined;
}

/**
 * type for the custom item toolbar item
 */
export interface CustomItemProps extends ToolbarItemBase {
  element: JSX.Element;
}

export const CustomToolbarItem: React.FC<CustomItemProps> = ({
  element
}: CustomItemProps): ToolbarItemElement => {
  return element;
};
