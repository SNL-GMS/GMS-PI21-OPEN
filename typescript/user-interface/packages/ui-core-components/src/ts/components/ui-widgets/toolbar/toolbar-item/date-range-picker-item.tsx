import type { DateTimeFormat } from '@gms/common-util';
import React from 'react';

import type { DateRangePickerTypes } from '../../date-range-picker';
import { DateRangePicker } from '../../date-range-picker';
import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isDateRangePickerToolbarItem(
  object: unknown
): object is DateRangePickerToolbarItemProps {
  return (
    (object as DateRangePickerToolbarItemProps).startTimeMs !== undefined &&
    (object as DateRangePickerToolbarItemProps).endTimeMs !== undefined &&
    (object as DateRangePickerToolbarItemProps).format !== undefined
  );
}

/**
 * Properties to pass to the {@link DateRangePickerToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface DateRangePickerToolbarItemProps extends ToolbarItemBase {
  /** start time - in milliseconds post Jan 1, 1970 */
  startTimeMs: number;

  /** end time - in milliseconds post Jan 1, 1970 */
  endTimeMs: number;

  /** format - @see DateTimeFormat */
  format: DateTimeFormat;

  /** duration - @see DateRangePickerTypes.DurationOption */
  durations?: DateRangePickerTypes.DurationOption[];

  /** Minimum time for the date picker to start - in milliseconds */
  minStartTimeMs?: number;

  /** Maximum time for the date picker to end - in milliseconds */
  maxEndTimeMs?: number;

  /** callback when the date picker value changes */
  onChange(startTimeMs: number, endTimeMs: number);

  /** callback when the date picker's apply button is clicked */
  onApplyButton(startTimeMs: number, endTimeMs: number);
}

/**
 * Represents a date picker item used within a toolbar
 *
 * @param dateRangePicker the datePickerItem to display {@link DateRangePickerItem}
 */
// eslint-disable-next-line react/function-component-definition
export const DateRangePickerToolbarItem: React.FC<DateRangePickerToolbarItemProps> = ({
  startTimeMs: startMs,
  endTimeMs: endMs,
  format,
  durations,
  minStartTimeMs,
  maxEndTimeMs,
  onChange,
  onApplyButton,
  style
}: DateRangePickerToolbarItemProps): ToolbarItemElement => {
  const handleNewInterval = (startTimeMs: number, endTimeMs: number) => {
    onChange(startTimeMs, endTimeMs);
  };

  const handleApply = (startTimeMs: number, endTimeMs: number) => {
    onApplyButton(startTimeMs, endTimeMs);
  };

  return (
    <div className="toolbar-button--capsule" style={style ?? {}}>
      <DateRangePicker
        startTimeMs={startMs}
        endTimeMs={endMs}
        format={format}
        durations={durations}
        minStartTimeMs={minStartTimeMs}
        maxEndTimeMs={maxEndTimeMs}
        onNewInterval={handleNewInterval}
        onApply={handleApply}
      />
    </div>
  );
};
