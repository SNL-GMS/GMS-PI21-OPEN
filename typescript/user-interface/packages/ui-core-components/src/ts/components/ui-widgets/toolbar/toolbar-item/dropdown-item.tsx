import React from 'react';

import { DropDown } from '../../drop-down';
import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isDropdownToolbarItem(object: unknown): object is DropdownToolbarItemProps {
  return (object as DropdownToolbarItemProps).dropdownOptions !== undefined;
}

/**
 * Properties to pass to the {@link DropdownToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface DropdownToolbarItemProps extends ToolbarItemBase {
  /** list of dropdown options that are selectable */
  dropdownOptions: any;

  /** list of text displayable options - one per dropdown option */
  dropdownText?: any;

  /** list of disabled options that should be disabled */
  disabledDropdownOptions?: any;

  /** item that is initially selected */
  value: any;

  custom?: boolean;

  /** should the label be displayed in front of the item list */
  displayLabel?: boolean;

  /** callback when the value of the list changes */
  onChange(value: any);
}

/**
 * Represents a group of static items to display/select within a toolbar
 *
 * @param dropdownItem the dropdownItem to display {@link DropdownItem}
 */
// eslint-disable-next-line react/function-component-definition
export const DropdownToolbarItem: React.FC<DropdownToolbarItemProps> = ({
  dropdownOptions,
  dropdownText,
  disabledDropdownOptions,
  value: dropdownValue,
  custom,
  displayLabel,
  onChange,
  style,
  disabled,
  widthPx,
  tooltip,
  label,
  cyData
}: DropdownToolbarItemProps): ToolbarItemElement => {
  return (
    <div style={style ?? {}}>
      <DropDown
        onMaybeValue={onChange}
        value={dropdownValue}
        custom={custom}
        dropDownItems={dropdownOptions}
        dropdownText={dropdownText}
        disabledDropdownOptions={disabledDropdownOptions}
        disabled={disabled}
        widthPx={widthPx}
        title={tooltip}
        data-cy={cyData}
        displayLabel={displayLabel}
        label={label}
      />
    </div>
  );
};
