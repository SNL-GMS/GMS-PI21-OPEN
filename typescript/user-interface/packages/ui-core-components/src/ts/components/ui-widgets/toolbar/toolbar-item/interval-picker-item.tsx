import React from 'react';

import { IntervalPicker } from '../../interval-picker';
import type { ToolbarItemBase, ToolbarItemElement } from '../types';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isIntervalPickerToolbarItem(
  object: unknown
): object is IntervalPickerToolbarItemProps {
  return (
    (object as IntervalPickerToolbarItemProps).startDate !== undefined &&
    (object as IntervalPickerToolbarItemProps).endDate !== undefined &&
    (object as IntervalPickerToolbarItemProps).defaultIntervalInHours !== undefined
  );
}

/**
 * Properties to pass to the {@link IntervalPickerToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface IntervalPickerToolbarItemProps extends ToolbarItemBase {
  /** start date for the interval */
  startDate: Date;

  /** end date for the interval */
  endDate: Date;

  /** default interval time - in hours */
  defaultIntervalInHours: number;

  /** display in short format - default is no */
  shortFormat?: boolean;

  /** callback when the interval is changed/selected */
  onChange(startDate: Date, endDate: Date);

  /** callback when the interval's apply button is clicked */
  onApplyButton(startDate: Date, endDate: Date);
}

/**
 * Represents an interval selector used within a toolbar
 *
 * @param intervalPickerItem the intervalPickerItem to display {@link IntervalPickerToolbarItem}
 */
// eslint-disable-next-line react/function-component-definition
export const IntervalPickerToolbarItem: React.FC<IntervalPickerToolbarItemProps> = ({
  startDate: start,
  endDate: end,
  defaultIntervalInHours,
  shortFormat,
  onChange,
  onApplyButton,
  style,
  cyData
}: IntervalPickerToolbarItemProps): ToolbarItemElement => {
  const handleNewInterval = (startDate: Date, endDate: Date) => {
    onChange(startDate, endDate);
  };

  const handleApply = (startDate: Date, endDate: Date) => {
    onApplyButton(startDate, endDate);
  };

  return (
    <div className="toolbar-button--capsule" style={style ?? {}}>
      <IntervalPicker
        shortFormat={shortFormat}
        startDate={start}
        endDate={end}
        onNewInterval={handleNewInterval}
        onApply={handleApply}
        defaultIntervalInHours={defaultIntervalInHours}
        data-cy={cyData}
      />
    </div>
  );
};
