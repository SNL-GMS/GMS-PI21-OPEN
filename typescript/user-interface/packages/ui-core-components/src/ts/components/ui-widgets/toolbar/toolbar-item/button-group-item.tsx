import React from 'react';

import type { ToolbarItemBase, ToolbarItemElement } from '../types';
import type { ButtonToolbarItemProps } from './button-item';
import { ButtonToolbarItem } from './button-item';

/**
 * Type guard, for use when rendering overflow menu items.
 */
export function isButtonGroupToolbarItem(object: unknown): object is ButtonGroupToolbarProps {
  return (object as ButtonGroupToolbarProps).buttons !== undefined;
}

/**
 * Properties to pass to the {@link ButtonGroupToolbarItem}
 *
 * @see {@link ToolbarItemBase} for base properties.
 */
export interface ButtonGroupToolbarProps extends ToolbarItemBase {
  /** Array of props for {@link ButtonToolbarItem} items to be rendered. */
  buttons: ButtonToolbarItemProps[];
}

/**
 * Represents a group of buttons used within a toolbar
 *
 * @param groupItem Properties to pass to the ButtonGroup {@link ButtonGroupToolbarProps}
 */
// eslint-disable-next-line react/function-component-definition
export const ButtonGroupToolbarItem: React.FC<ButtonGroupToolbarProps> = ({
  buttons,
  style,
  cyData
}: ButtonGroupToolbarProps): ToolbarItemElement => {
  const indexOfLastButton = buttons.length - 1;
  return (
    <div className="toolbar-button-group" data-cy={cyData} style={style ?? {}}>
      {buttons.map((button, index) => {
        return (
          <ButtonToolbarItem
            key={button.key}
            widthPx={button.widthPx}
            onlyShowIcon={button.onlyShowIcon}
            style={button.style}
            disabled={button.disabled}
            label={button.label}
            labelRight={button.labelRight}
            tooltip={button.tooltip}
            icon={button.icon}
            onButtonClick={() => button.onButtonClick()}
            onMouseEnter={button.onMouseEnter ? () => button.onMouseEnter() : undefined}
            onMouseOut={button.onMouseOut ? () => button.onMouseOut() : undefined}
            marginRight={index !== indexOfLastButton ? 2 : 0}
            cyData={button.cyData}
          />
        );
      })}
    </div>
  );
};
