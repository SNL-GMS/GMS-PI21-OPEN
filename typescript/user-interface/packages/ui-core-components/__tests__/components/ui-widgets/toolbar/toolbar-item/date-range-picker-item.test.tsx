import { render } from '@testing-library/react';
import React from 'react';

import { DateRangePickerToolbarItem } from '../../../../../src/ts/components/ui-widgets/toolbar/toolbar-item/date-range-picker-item';

// set up window alert and open map so we don't see errors
window.alert = jest.fn();
window.open = jest.fn();

describe('DateRangePickerToolbarItem', () => {
  test('DateRangePickerToolbarItem renders directly', () => {
    const { container } = render(
      <DateRangePickerToolbarItem
        key="datepicker"
        startTimeMs={100}
        endTimeMs={1000}
        onChange={jest.fn()}
        format={undefined}
        onApplyButton={jest.fn()}
        tooltip="Hello Date Range"
        hasIssue={false}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
