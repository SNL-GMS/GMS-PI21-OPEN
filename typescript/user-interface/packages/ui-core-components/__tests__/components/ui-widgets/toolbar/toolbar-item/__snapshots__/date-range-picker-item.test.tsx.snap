// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangePickerToolbarItem DateRangePickerToolbarItem renders directly 1`] = `
<div>
  <div
    class="toolbar-button--capsule"
  >
    <div
      class="date-range-picker"
      id="date-range-picker-id-1"
    >
      <label
        class="bp4-label bp4-label bp4-inline date-range-picker__label"
      >
        Time Range: 1970-01-01 00:00 to 1970-01-01 00:00
      </label>
      <div>
        <button
          class="bp4-button date-range-picker__apply-button"
          data-cy="date-range-picker-edit"
          type="button"
        >
          <span
            class="bp4-button-text"
          >
            Edit
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
